Document addition of namespace labels for pods needing elevated privileges (#1706)
spectrocloud/librarium: docs/docs-content/troubleshooting/cluster-deployment.md

6. Check stdout for errors. You can also open a support ticket. Visit our [support page](http://support.spectrocloud.io/).

## Deployment Violates Pod Security

> In cluster profiles running Kubernetes 1.25 or later, some packs contain Pods that need elevated privileges during Pod creation. 

Reviewer: You may want to start by describing the scenario. Maybe something like the following.

> In cluster profiles running Kubernetes 1.25 or later, some packs contain Pods that need elevated privileges during Pod creation. 

Reviewer: ```suggestion
Cluster deployment fails with the following message. 
```

Reviewer: Include the message here.

> When you try to deploy a cluster with such packs, you get an error message that looks like the following:

Reviewer: ```suggestion
This can happen when the cluster profile uses Kubernetes 1.25 or later and indicates the pod needs elevated privileges to be created. 
```

```
Error creating: pods <name of pod> is forbidden: violates PodSecurity "baseline:v<k8s version>": non-default capabilities …
```

### Debug Steps

You can change the Pod Security Standards of the namepace where the Pod is being created to address this issue. 

> 

Reviewer: You can find the style we use for basic login to Palette [Update a Cluster Profile].(https://docs.spectrocloud.com/profiles/cluster-profiles/modify-cluster-profiles/update-cluster-profile).

![CleanShot 2023-10-24 at 16 15 16](https://github.com/spectrocloud/librarium/assets/117382432/91bb8aef-9d57-446d-8684-738af6f5378a)

1. Log in to **Palette** and navigate to the **Profiles** tab. 
2. Select the profile you are trying to deploy the cluster with and choose the layer that represents your pack. 

> 3. In the YAML file for your pack, under the `pack` field, add a subfield `namespaceLabels`.

Reviewer: ```suggestion
3. In the `pack` section of the Kubernetes YAML file, add a subsection called `namespaceLabels`.
```

Reviewer: Lenny, as you and I discussed, it might be _other_ packs in combination with Kubernetes packs (1.25 and higher).

6. Check stdout for errors. You can also open a support ticket. Visit our [support page](http://support.spectrocloud.io/).

## Deployment Violates Pod Security
Cluster deployment fails with the following message. 

```
Error creating: pods <name of pod> is forbidden: violates PodSecurity "baseline:v<k8s version>": non-default capabilities …
```

> This can happen when the cluster profile uses Kubernetes 1.25 or later and also includes packs that create Pods requiring elevated privileges . 

Reviewer: We don't typically initial cap resources.

### Debug Steps

> You can change the Pod Security Standards of the namepace where the Pod is being created to address this issue. 

Reviewer: ```suggestion
To address this issue, you can change the Pod Security Standards of the namespace where the pod is being created. 
```

Reviewer: It's up to you how to start the sentence - to me, it just sounds better to start with "To address this issue..."

1. Log in to [Palette](https://console.spectrocloud.com).

2. Navigate to the left **Main Menu** and click on **Profiles**. 

> 3. Select the profile you are trying to deploy the cluster with and choose the layer that represents your pack. 

Reviewer: You're on the right track with our terminology to "choose the layer". I just wonder if there's a way to be more specific about "that represents your pack".

Possibly: "...and choose the layer that configures a pod."

> 3. Select the profile you are trying to deploy the cluster with and choose the layer that represents your pack. 

Reviewer: We could break out step 3 into two separate steps - something like:

3. Select the profile you are using to deploy the cluster. Palette displays the profile stack and details.
4. Click on the layer in the profile stack that contains the pod configuration.

4. In the YAML file for that pack, under the `pack` field, add a subfield `namespaceLabels` if it doesn't already exist.

> 5. In the `namespaceLabels` field, add a subfield with the name of your namespace as the key and add `pod-security.kubernetes.io/enforce=privileged,pod-security.kubernetes.io/enforce-version=v<k8s_version>` as its value. Replace `<k8s_version>` with the version of Kubernetes that runs on your cluster. 

Reviewer: ```suggestion
5. In the `namespaceLabels` section, add a subsection with the name of your namespace as the key and add `pod-security.kubernetes.io/enforce=privileged,pod-security.kubernetes.io/enforce-version=v<k8s_version>` as its value. Replace `<k8s_version>` with the version of Kubernetes that runs on your cluster. 
```

   - If a key matching your namespace already exists here, add the labels to the value corresponding to that key.

>    - For example, if you the pack creates a namespace called `monitoring`, add the labels to the `monitoring` namespace:

Reviewer: Missing word. Should it be "For example, if your pack creates a namespace called `monitoring`, add the labels to the `monitoring` namespace as shown in the example below.

>    ```yaml

Reviewer: ```suggestion
   
   ```yaml
```

Reviewer: Added a line between the last bullet and the yaml example.

   pack:
    namespace: "monitoring"

    namespaceLabels:
        "monitoring": "org=spectro,team=dev,pod-security.kubernetes.io/enforce=privileged,pod-security.kubernetes.io/enforce-version=v1.28"
   ```

> 

Reviewer: Let's tab the example over so it aligns with the steps.

```

This can happen when the cluster profile uses Kubernetes 1.25 or later and also includes packs that create pods requiring elevated privileges. 

### Debug Steps

To address this issue, you can change the Pod Security Standards of the namespace where the pod is being created. 

1. Log in to [Palette](https://console.spectrocloud.com).

1. Navigate to the left **Main Menu** and click on **Profiles**. 

1. Select the profile you are using to deploy the cluster. Palette displays the profile stack and details.
Click on the layer in the profile stack that contains the pack configuration.

1. In the pack's YAML file, add a subfield in the `pack` section called `namespaceLabels` if it does not already exist.

1. In the `namespaceLabels` section, add a subsection with the name of your namespace as the key and add `pod-security.kubernetes.io/enforce=privileged,pod-security.kubernetes.io/enforce-version=v<k8s_version>` as its value. Replace `<k8s_version>` with the version of Kubernetes that runs on your cluster. 
   - If a key matching your namespace already exists here, add the labels to the value corresponding to that key. 

:::tip

> If your pack creates multiple namespaces, and you are not sure which namespaces need the elevated privileges, you can [access the cluster with the kubectl CLI](https://docs.spectrocloud.com/clusters/cluster-management/palette-webctl/#access-cluster-with-cli) and use [`kubectl get pods`](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#get) to find out which pods are failing at creation in which namespaces. We recommend that you only apply the labels to namespaces where pods are failing to be created. 

Reviewer: We usually link to docs with a sentence like the suggested ones. We also need to use relative paths with Docusaurus. Links won't include https:// unless your linking outside of Spectro docs. And they won't include docs.spectrocloud.

6. Check stdout for errors. You can also open a support ticket. Visit our [support page](http://support.spectrocloud.io/).

## Deployment Violates Pod Security

> Cluster deployment fails with the following message. 

Reviewer 2: ```suggestion
Cluster deployment fails with the following message. 2222
```
